Alethe proofs rejected with rule mod_simplify

The alethe proofs for [chc-LIA_366](https://github.com/chc-comp/chc-comp23-benchmarks/blob/main/LIA-lin/chc-LIA-Lin_366.smt2.gz) and [chc-LIA_379](https://github.com/chc-comp/chc-comp23-benchmarks/blob/main/LIA-lin/chc-LIA-Lin_379.smt2.gz), from CHC-COMP 2023, are being rejected by [carcara](https://github.com/ufmg-smite/carcara).

Production command: `golem --logic QF_LIA --print-witness --proof-format alethe [file.smt2]`
Checking command: `carcara check --allow-int-real-subtyping --expand-let-bindings [proof.alethe] [file.smt2]`

Checking output for chc-LIA_366:

```
[ERROR] checking failed on step 't77427' with rule 'mod_simplify': expected term '1' to be numerical constant 9.0
invalid
```

Checking output for chc-LIA_379:

```
[ERROR] checking failed on step 't6544' with rule 'mod_simplify': expected term '26' to be numerical constant 63.0
invalid
```

--- a/src/proofs/Term.cc
+++ b/src/proofs/Term.cc
@@ -484,13 +484,24 @@ std::shared_ptr<Term> Op::operate() const {
             return args[2];
         }
     } else if (operation == "mod") {
-        FastRational result = firstTerm % secondTerm;
-        if (result < 0) {
-            result *= -1;
-            return std::make_shared<Terminal>("(- " + result.get_str() + ")", Term::INT);
-        } else {
-            return std::make_shared<Terminal>(result.get_str(), Term::INT);
-        }
+        // NOTE: The semantics of modulo operation in OpenSMT's FastRationals is different from the semantics defined by
+        // SMT-LIB. We must use the computation of the remainder according to SMT-LIB.
+        // See notes in // https://smtlib.cs.uiowa.edu/theories-Ints.shtml
+
+        // "Regardless of sign of m,
+        //  when n is positive, (div m n) is the floor of the rational number m/n;
+        //  when n is negative, (div m n) is the ceiling of m/n."
+
+        // Remainder is then always a positive number r such that m = n * q + r, r < abs(n)
+        auto smtlib_modulo = [](auto const & m, auto const & n) -> FastRational {
+            auto ratio = m / n;
+            auto q = n > 0 ? ratio.floor() : ratio.ceil();
+            auto r = m - n * q;
+            assert(r.isInteger() and r.sign() >= 0 and r < abs(n));
+            return r;
+        };
+        FastRational result = smtlib_modulo(firstTerm, secondTerm);
+        return std::make_shared<Terminal>(result.get_str(), Term::INT);
     } else if (operation == "div") {
         FastRational result = firstTerm / secondTerm;
         if (result < 0) {

--- a/test/prooftesting/negative_mod.smt2
+++ b/test/prooftesting/negative_mod.smt2
@@ -1,0 +1,11 @@
+(set-logic HORN)
+(declare-fun Q (Int) Bool)
+(assert (forall ((x Int)) (=> (and (< x 0) (= (mod x 3) 1)) (Q x))
+))
+
+
+(assert (forall ((x Int))
+    (=> (Q x) false)
+))
+
+(check-sat)

